rmw_opendds_cpp/src/rmw_publish.cpp

```diff
@@ -26,12 +26,26 @@
 
 #include <ace/Message_Block.h>
 
+bool set_key(OpenDDSStaticSerializedData & instance, DDS::DataWriter * dds_data_writer) {
+  OpenDDS::DCPS::DataWriterImpl* dwImpl = dynamic_cast<OpenDDS::DCPS::DataWriterImpl*>(dds_data_writer);
+  if (!dwImpl) {
+    RMW_SET_ERROR_MSG("failed to cast dds_data_writer to DataWriterImpl");
+    return false;
+  }
+  OpenDDS::DCPS::Message_Block_Ptr key(new ACE_Message_Block(KEY_HASH_LENGTH_16, ACE_Message_Block::MB_DATA, 0, 0, 0, 0));
+  OpenDDS::DCPS::Serializer ser(key.get());
+  ser << dwImpl->get_publication_id();
+  instance.serialized_key.length(KEY_HASH_LENGTH_16);
+  std::memcpy(instance.key_hash, key.get(), KEY_HASH_LENGTH_16);
+  std::memcpy(instance.serialized_key.get_buffer(), key.get(), KEY_HASH_LENGTH_16);
```

Comment:
Just a minor comment that swapping lines 39 and 40 might make the code easier to follow since lines 38 and 40 are both modifying `instance.serialized_key`.

Reply:
Yeah, I see what you mean. The primary intension is to keep the order of the declaration: 
@topic struct OpenDDSStaticSerializedData {
  @key OctetArray16       key_hash;
  DDS::OctetSeq  serialized_key;
  DDS::OctetSeq  serialized_data;
};

To that end, the code is like this:
OpenDDS::DCPS::RepoId id = dwImpl->get_publication_id();
std::memcpy(instance.key_hash, &id, KEY_HASH_LENGTH_16);
instance.serialized_key.length(KEY_HASH_LENGTH_16);
std::memcpy(instance.serialized_key.get_buffer(), &id, KEY_HASH_LENGTH_16);

If I don't factor out this function, the above code would be immediately followed by these two lines to complete the initialization:
instance.serialized_data.length(static_cast<CORBA::ULong>(cdr_stream->buffer_length));
std::memcpy(instance.serialized_data.get_buffer(), cdr_stream->buffer, cdr_stream->buffer_length);

Reply:
Yes, I agree.